src/Microsoft.OData.Core/UriParser/CustomUriFunctionsStore.cs

```diff
@@ -0,0 +1,234 @@
+        /// <param name="functionSignatures">
+        /// When this method returns, contains the list of function signatures if found; otherwise, null.
+        /// </param>
+        /// <returns><c>true</c> if one or more overloads were found; otherwise, <c>false</c>.</returns>
+        public bool TryGet(string functionName, bool ignoreCase, out IReadOnlyList<FunctionSignatureWithReturnType> functionSignatures)
+        {
+            ExceptionUtils.CheckArgumentStringNotNullOrEmpty(functionName, "functionName");
+
+            functionSignatures = null;
+
+            if (!ignoreCase)
+            {
+                if (this.map.TryGetValue(functionName, out List<FunctionSignatureWithReturnType> signatures))
+                {
+                    functionSignatures = signatures.AsReadOnly();
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            StringComparer comparer = ignoreCase ? StringComparer.OrdinalIgnoreCase : StringComparer.Ordinal;
+
+            foreach (KeyValuePair<string, List<FunctionSignatureWithReturnType>> kvp in this.map)
+            {
+                if (comparer.Equals(kvp.Key, functionName))
+                {
+                    functionSignatures = kvp.Value.AsReadOnly();
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Adds a custom URI function signature for the specified function name.
+        /// If the function name already exists, the signature is added as an overload if it is not already present.
+        /// </summary>
+        /// <param name="functionName">The name of the custom function.</param>
+        /// <param name="functionSignature">The function signature to add.</param>
+        public void Add(string functionName, FunctionSignatureWithReturnType functionSignature)
+        {
+            ExceptionUtils.CheckArgumentStringNotNullOrEmpty(functionName, "functionName");
+            ExceptionUtils.CheckArgumentNotNull(functionSignature, "functionSignature");
+
+            CompareExchangeLoop(
+                ref this.map,
+                snapshot =>
+                {
+                    if (!snapshot.TryGetValue(functionName, out List<FunctionSignatureWithReturnType> existingSignatures))
+                    {
+                        return snapshot.Add(functionName, new List<FunctionSignatureWithReturnType> { functionSignature });
+                    }
+                    else
+                    {
+                        List<FunctionSignatureWithReturnType> updatedSignatures = new List<FunctionSignatureWithReturnType>(existingSignatures);
+                        if (!updatedSignatures.Contains(functionSignature))
+                        {
+                            updatedSignatures.Add(functionSignature);
+                        }
+
+                        return snapshot.SetItem(functionName, updatedSignatures);

```

Comment:
perf: To avoid resizing, we could set a capacity that accounts for the item being added, otherwise, we'll always double the list capacity since creating a list based on the existing list will initialize the capacity to the input size, and it would get doubled when we add one more item to it: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs,75

```suggestion
                        if (!existingSignatures.Contains(functionSignature)
                        {
                            List<FunctionSignatureWithReturnType> updatedSignatures = new(existingSignatures.Count + 1);
                            updatedSignatures.AddRange(existingSignatures);
                            updateSignatures.Add(functionSignature);
                            return snapshot.SetItem(functionName, updatedSignature);
                        }

                         // no update
                        return snapshot;
}
```